[Banner] Use autolayout internally for BannerView.

fluxminds opened the pull request:
Major changes for review:

- Remove `MDCBannerViewLayout` class to reduce confusion on `layout` and simplify code structure.
- Use Autolayout to layout subviews in MDCBannerView to simplify the layout calculation logic

Note:
- This class is intended to only support iOS9+ following up the design meeting, though support for iOS8 could be recovered.
- This PR is a follow-up to https://github.com/material-components/material-components-ios/pull/6963. The API of this MDCBannerView has been reviewed. This PR serves as implementation details.
- This PR is also a follow-up PR split from https://github.com/material-components/material-components-ios/pull/6830. The original version was approved, this is a much simpler version with less code.

Constraints Screenshot:
(1) Single Line:
![SingleLineMode](https://user-images.githubusercontent.com/8836258/55303898-dae58780-5416-11e9-9a64-b06330368c07.png)
(2) MultiLine Stacked Buttons
![MultiLineWithStackedButton](https://user-images.githubusercontent.com/8836258/55303904-e46eef80-5416-11e9-956e-c38c06f97e8c.png)
(3) MultiLine Aligned Buttons
![MultiLineAlignedButtonMode](https://user-images.githubusercontent.com/8836258/55303914-f05ab180-5416-11e9-8e42-4fa3a5ea8cd8.png)

Commit:
[Banner] Use autolayout internally for BannerView.

material-automation commented:
bazel detected changes to the following targets:

```
//components/Banner:ObjcExamples
```

<!-- identifier: affected-targets -->

randallli commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -14,24 +14,39 @@
+#import "MaterialTypography.h"
+
+typedef NS_ENUM(NSInteger, MDCBannerViewLayoutStyle) {
+  MDCBannerViewLayoutSingleLineStyle = 0,              // All elements lays on the same line
```
copyedit of comment: All elements on the same line

fluxminds replied:
Addressed in 070341d4e3cb361c899765e68b6be6ac8e7266aa.

randallli commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +69,312 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
 
-- (void)setText:(NSString *)text {
-  self.textLabel.text = text;
+- (BOOL)isImageSet {
```
rename function: hasImage

it is more idiomatic to objective c and UIKit

fluxminds replied:
Addressed in 070341d.

randallli commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +69,312 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+      [self.leadingButton sizeToFit];
+      CGFloat maximumHeight =
+          MAX(CGRectGetHeight(self.textLabel.frame), CGRectGetHeight(self.leadingButton.frame));
+      if (self.isImageSet) {
```
lets change this to be [self isImageSet]
. notations is for properties.

fluxminds replied:
It is a `readonly` property, so I kept the dot notation for now.

randallli commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +69,312 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  _trailingButton = trailingButton;
 }
 
 #pragma mark - Property Setter and Getter
```
rename: Helper methods
because this pragma is no longer accurate.

fluxminds replied:
Changed to `Property Getter` in 070341d4e3cb361c899765e68b6be6ac8e7266aa because `hasImage` is a `readonly` property.

randallli commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +69,312 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  switch (style) {
+    case MDCBannerViewLayoutSingleLineStyle: {
+      frameHeight += kTopPaddingSmall + kBottomPadding;
+      [self.textLabel sizeToFit];
```
it is a odd to call sizeToFit from within a sizeThatFits. More often sizeThatFits calls more sizeThatFits of its subviews.

Think of it this way, sizeThatFits is asking the banner view what size would you be if you were told you had this much space. We do not expect the subviews (label and buttons) to change size when we ask this hypothetical question.

https://www.quora.com/Whats-the-difference-between-sizeToFit-sizeThatFits-and-sizeWithAttributes

fluxminds replied:
Thanks for suggestion. I was using `sizeToFit` and then getting the frame, but it is right that changing the actual frame is unnecessary. 

Addressed in 070341d.

randallli commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +69,312 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  [super layoutSubviews];
+
+  if (![self isImageSet]) {
+    self.imageView.hidden = YES;
```
It is odd to turn this on when you layoutSubviews because it overrides what the client might have set. 

When will we this ever change the hidden property? The method call in the if clause seems to be checking the hidden property already.

fluxminds replied:
Yea. This was one part I felt unclean and odd too. The reason why I wanted to do was UIButton set `hidden` when `setImage:nil..` is called. So I want to keep the same behavior. 

I lean more towards using a KVO on `imageView.image`. Addressed in 070341d.

randallli commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +69,312 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  remainingWidth -= marginsPadding;
+  remainingWidth -= (kLeadingPadding + kTrailingPadding);
+  if (self.trailingButton.hidden) {
+    [self.leadingButton sizeToFit];
```
you should move things that are in both branches of an if else outside of the conditional entirely

fluxminds replied:
Addressed in 070341d.

Commit:
CRC changes: remove unclean layoutSubview override

jverkoey commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +70,330 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+      break;
+    }
+  }
+  return CGSizeMake(size.width, frameHeight);
```
I'm surprised that we still need custom logic in `sizeThatFits:` if we're supporting auto layout, are we able to convert this remaining logic to constraints?

fluxminds replied:
When `custom logic` is mentioned, do you mean the logic to calculate `style` or the entire logic here? For the entire logic, it is used to decided the view size based on given properties (label, image..in this case). For the `style` logic, it might not be easy and straightforward to do everything directly with constraints. For example, the spec requires the top padding to be 10 when all elements are able to fit on the same line while it to be 24 when element are not able to fit. So it is easier to update constraint based on this `style`. There might be a way to fully use constraints, but using some manual calculation here to get `style` would make it easier to set constraints.

Also because you mentioned `if we're supporting auto layout`, there is one thing worth noting here. `instrinsicContentSize` is not able to be implemented here because we don't have a `preferredContentWidth` property. So developer who uses auto layout needs to add specific constraints for its size.

jverkoey replied:
I think I understand. Though I'm still curious whether it's possible to make this "last mile" of layout considerations calculated via constraints as well, this is a small enough implementation detail that I won't block review of this PR on it.

jverkoey commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +70,330 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+      [self.buttonContainerView.bottomAnchor
+          constraintEqualToAnchor:self.layoutMarginsGuide.bottomAnchor
+                         constant:-kBottomPadding]
+          .active = YES;
```
I don't see it in this code, but I'd expect the existing constraints to be removed at some point or at least disabled. Did you find that wasn't necessary when the layout style changed?

fluxminds replied:
Good catch! That part was missing from this PR. I will work on adding it when refactoring this code block following up the comment above.

fluxminds replied:
Addressed in https://github.com/material-components/material-components-ios/commit/63130a82404872e1085ad6201c4100a5de31e84f#diff-9f546e78faa17de3d67395c6ef3629b2R306.

jverkoey commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -14,24 +14,40 @@
 #import "MaterialButtons.h"
+#import "MaterialTypography.h"
+
+typedef NS_ENUM(NSInteger, MDCBannerViewLayoutStyle) {
```
Suggestion: `MDCBannerViewLayoutMode`

This doesn't particularly have to do with style.

fluxminds replied:
Thanks for the suggestion. Addressed in a34769e62030086202b79c041505a0dd813563f7.

jverkoey commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +70,330 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  if (@available(iOS 9.0, *)) {
+    if (layoutStyle == MDCBannerViewLayoutSingleLineStyle) {
+      if (self.hasImage) {
+        [self.imageView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor
```
Is it possible/reasonable to initialize these constraints elsewhere and to then to have this method focus on enabling/disabling the necessary groups of constraints? I think that may simplify the logic here.

fluxminds replied:
Good idea. I will refactor this part and try to see if I can make it clean.

fluxminds replied:
Addressed in  63130a8

Commit:
CRC: rename style to mode.

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +69,330 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    }
+    layoutMode = [self isAbleToFitTextLabel:self.textLabel withWidthLimit:remainingWidth]
+                ? MDCBannerViewLayoutModeSingleLine
+                : MDCBannerViewLayoutModeMultiLineAlignedButton;
```
clang-format suggested the following change:


```
-                ? MDCBannerViewLayoutModeSingleLine
-                : MDCBannerViewLayoutModeMultiLineAlignedButton;
+                     ? MDCBannerViewLayoutModeSingleLine
+                     : MDCBannerViewLayoutModeMultiLineAlignedButton;
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +69,330 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    CGFloat buttonWidth = [self widthSumForButtons:@[ self.leadingButton, self.trailingButton ]];
+    remainingWidth -= buttonWidth;
+    layoutMode = (remainingWidth > 0) ? MDCBannerViewLayoutModeMultiLineAlignedButton
+                                 : MDCBannerViewLayoutModeMultiLineStackedButton;
```
clang-format suggested the following change:


```
-                                 : MDCBannerViewLayoutModeMultiLineStackedButton;
+                                      : MDCBannerViewLayoutModeMultiLineStackedButton;
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
CRC: initialize contraints in a seperate setup method.

Commit:
some missing fixes.

Commit:
fix format

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -14,24 +14,73 @@
+// Buttons constraints
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintLeading;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintLeadingWithTextLabel;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintLeadingWithTextLabelGreater;
```
clang-format suggested the following change:


```
-@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintLeadingWithTextLabel;
-@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintLeadingWithTextLabelGreater;
+@property(nonatomic, readwrite, strong)
+    NSLayoutConstraint *buttonContainerConstraintLeadingWithTextLabel;
+@property(nonatomic, readwrite, strong)
+    NSLayoutConstraint *buttonContainerConstraintLeadingWithTextLabelGreater;
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -14,24 +14,73 @@
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintTopWithMargin;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintTopWithImageViewGreater;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintTopWithTextLabel;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintTopWithTextLabelGreater;
```
clang-format suggested the following change:


```
-@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintTopWithImageViewGreater;
-@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintTopWithTextLabel;
-@property(nonatomic, readwrite, strong) NSLayoutConstraint *buttonContainerConstraintTopWithTextLabelGreater;
+@property(nonatomic, readwrite, strong)
+    NSLayoutConstraint *buttonContainerConstraintTopWithImageViewGreater;
+@property(nonatomic, readwrite, strong)
+    NSLayoutConstraint *buttonContainerConstraintTopWithTextLabel;
+@property(nonatomic, readwrite, strong)
+    NSLayoutConstraint *buttonContainerConstraintTopWithTextLabelGreater;
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -14,24 +14,73 @@
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *leadingButtonConstraintTop;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *leadingButtonConstraintTrailing;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *leadingButtonConstraintCenterY;
+@property(nonatomic, readwrite, strong) NSLayoutConstraint *leadingButtonConstraintTrailingWithTrailingButton;
```
clang-format suggested the following change:


```
-@property(nonatomic, readwrite, strong) NSLayoutConstraint *leadingButtonConstraintTrailingWithTrailingButton;
+@property(nonatomic, readwrite, strong)
+    NSLayoutConstraint *leadingButtonConstraintTrailingWithTrailingButton;
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -14,24 +14,73 @@
+
 
-@property(nonatomic, readwrite, strong) MDCBannerViewLayout *layout;
 
```
clang-format suggested the following change:


```
-
-
-
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +103,325 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    self.textLabelConstraintLeadingWithMargin = [self.textLabel.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
+
+    self.buttonContainerConstraintLeading = [self.buttonContainerView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
+    self.buttonContainerConstraintTrailing = [self.textLabel.trailingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor constant:-kTrailingPadding];
```
clang-format suggested the following change:


```
-    self.imageViewConstraintLeading = [self.imageView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
-    self.imageViewConstraintTopSmall = [self.imageView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor constant:kTopPaddingSmall];
-    self.imageViewConstraintTopLarge = [self.imageView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor constant:kTopPaddingLarge];
-    self.imageViewConstraintBottom = [self.imageView.bottomAnchor constraintEqualToAnchor:self.layoutMarginsGuide.bottomAnchor constant:-kBottomPadding];
-
-    self.textLabelConstraintTop = [self.textLabel.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor constant:kTopPaddingLarge];
-    self.textLabelConstraintCenterY = [self.textLabel.centerYAnchor constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
-    self.textLabelConstraintTrailing = [self.textLabel.trailingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor constant:-kTrailingPadding];
-    self.textLabelConstraintLeadingWithImage = [self.textLabel.leadingAnchor constraintEqualToAnchor:self.imageView.trailingAnchor constant:kSpaceBetweenIconImageAndTextLabel];
-    self.textLabelConstraintLeadingWithMargin = [self.textLabel.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
-
-    self.buttonContainerConstraintLeading = [self.buttonContainerView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
-    self.buttonContainerConstraintTrailing = [self.textLabel.trailingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor constant:-kTrailingPadding];
-    self.buttonContainerConstraintBottom = [self.buttonContainerView.bottomAnchor constraintEqualToAnchor:self.layoutMarginsGuide.bottomAnchor constant:-kBottomPadding];
-    self.buttonContainerConstraintLeadingWithTextLabel = [self.buttonContainerView.leadingAnchor constraintEqualToAnchor:self.textLabel.trailingAnchor constant:kHorizontalSpaceBetweenTextLabelAndButton];
+    self.imageViewConstraintLeading =
+        [self.imageView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor
+                                                     constant:kLeadingPadding];
+    self.imageViewConstraintTopSmall =
+        [self.imageView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor
+                                                 constant:kTopPaddingSmall];
+    self.imageViewConstraintTopLarge =
+        [self.imageView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor
+                                                 constant:kTopPaddingLarge];
+    self.imageViewConstraintBottom =
+        [self.imageView.bottomAnchor constraintEqualToAnchor:self.layoutMarginsGuide.bottomAnchor
+                                                    constant:-kBottomPadding];
+
+    self.textLabelConstraintTop =
+        [self.textLabel.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor
+                                                 constant:kTopPaddingLarge];
+    self.textLabelConstraintCenterY = [self.textLabel.centerYAnchor
+        constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
+    self.textLabelConstraintTrailing = [self.textLabel.trailingAnchor
+        constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor
+                       constant:-kTrailingPadding];
+    self.textLabelConstraintLeadingWithImage =
+        [self.textLabel.leadingAnchor constraintEqualToAnchor:self.imageView.trailingAnchor
+                                                     constant:kSpaceBetweenIconImageAndTextLabel];
+    self.textLabelConstraintLeadingWithMargin =
+        [self.textLabel.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor
+                                                     constant:kLeadingPadding];
+
+    self.buttonContainerConstraintLeading = [self.buttonContainerView.leadingAnchor
+        constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor
+                       constant:kLeadingPadding];
+    self.buttonContainerConstraintTrailing = [self.textLabel.trailingAnchor
+        constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor
+                       constant:-kTrailingPadding];
+    self.buttonContainerConstraintBottom = [self.buttonContainerView.bottomAnchor
+        constraintEqualToAnchor:self.layoutMarginsGuide.bottomAnchor
+                       constant:-kBottomPadding];
+    self.buttonContainerConstraintLeadingWithTextLabel = [self.buttonContainerView.leadingAnchor
+        constraintEqualToAnchor:self.textLabel.trailingAnchor
+                       constant:kHorizontalSpaceBetweenTextLabelAndButton];
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +103,325 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    self.buttonContainerConstraintLeadingWithTextLabel = [self.buttonContainerView.leadingAnchor constraintEqualToAnchor:self.textLabel.trailingAnchor constant:kHorizontalSpaceBetweenTextLabelAndButton];
+    self.buttonContainerConstraintLeadingWithTextLabel.priority = UILayoutPriorityDefaultHigh;
+    self.buttonContainerConstraintLeadingWithTextLabelGreater = [self.buttonContainerView.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.textLabel.trailingAnchor constant:kHorizontalSpaceBetweenTextLabelAndButton];
+    self.buttonContainerConstraintTopWithMargin = [self.buttonContainerView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor constant:kTopPaddingSmall];
```
clang-format suggested the following change:


```
-    self.buttonContainerConstraintLeadingWithTextLabelGreater = [self.buttonContainerView.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.textLabel.trailingAnchor constant:kHorizontalSpaceBetweenTextLabelAndButton];
-    self.buttonContainerConstraintTopWithMargin = [self.buttonContainerView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor constant:kTopPaddingSmall];
-    self.buttonContainerConstraintTopWithImageViewGreater = [self.buttonContainerView.topAnchor constraintGreaterThanOrEqualToAnchor:self.imageView.bottomAnchor constant:kVerticalSpaceBetweenButtonAndTextLabel];
-    self.buttonContainerConstraintTopWithTextLabel = [self.buttonContainerView.topAnchor constraintEqualToAnchor:self.textLabel.bottomAnchor constant:kVerticalSpaceBetweenButtonAndTextLabel];
+    self.buttonContainerConstraintLeadingWithTextLabelGreater =
+        [self.buttonContainerView.leadingAnchor
+            constraintGreaterThanOrEqualToAnchor:self.textLabel.trailingAnchor
+                                        constant:kHorizontalSpaceBetweenTextLabelAndButton];
+    self.buttonContainerConstraintTopWithMargin = [self.buttonContainerView.topAnchor
+        constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor
+                       constant:kTopPaddingSmall];
+    self.buttonContainerConstraintTopWithImageViewGreater = [self.buttonContainerView.topAnchor
+        constraintGreaterThanOrEqualToAnchor:self.imageView.bottomAnchor
+                                    constant:kVerticalSpaceBetweenButtonAndTextLabel];
+    self.buttonContainerConstraintTopWithTextLabel = [self.buttonContainerView.topAnchor
+        constraintEqualToAnchor:self.textLabel.bottomAnchor
+                       constant:kVerticalSpaceBetweenButtonAndTextLabel];
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +103,325 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    self.leadingButtonConstraintTrailing = [self.trailingButton.trailingAnchor constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
+    self.leadingButtonConstraintCenterY = [self.leadingButton.centerYAnchor constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
+    self.leadingButtonConstraintTrailingWithTrailingButton = [self.leadingButton.trailingAnchor constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
+    self.trailingButtonConstraintBottom = [self.trailingButton.bottomAnchor constraintEqualToAnchor:self.buttonContainerView.bottomAnchor];
```
clang-format suggested the following change:


```
-    self.buttonContainerConstraintTopWithTextLabelGreater = [self.buttonContainerView.topAnchor constraintGreaterThanOrEqualToAnchor:self.textLabel.bottomAnchor constant:kVerticalSpaceBetweenButtonAndTextLabel];
-
-    self.leadingButtonConstraintLeading = [self.leadingButton.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.buttonContainerView.leadingAnchor];
-    self.leadingButtonConstraintTop = [self.leadingButton.topAnchor constraintEqualToAnchor:self.buttonContainerView.topAnchor];
-    self.leadingButtonConstraintTrailing = [self.trailingButton.trailingAnchor constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
-    self.leadingButtonConstraintCenterY = [self.leadingButton.centerYAnchor constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
-    self.leadingButtonConstraintTrailingWithTrailingButton = [self.leadingButton.trailingAnchor constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
-    self.trailingButtonConstraintBottom = [self.trailingButton.bottomAnchor constraintEqualToAnchor:self.buttonContainerView.bottomAnchor];
-    self.trailingButtonConstraintTop = [self.trailingButton.topAnchor constraintEqualToAnchor:self.leadingButton.bottomAnchor constant:kButtonVerticalIntervalSpace];
-    self.trailingButtonConstraintTrailing = [self.trailingButton.trailingAnchor constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
+    self.buttonContainerConstraintTopWithTextLabelGreater = [self.buttonContainerView.topAnchor
+        constraintGreaterThanOrEqualToAnchor:self.textLabel.bottomAnchor
+                                    constant:kVerticalSpaceBetweenButtonAndTextLabel];
+
+    self.leadingButtonConstraintLeading = [self.leadingButton.leadingAnchor
+        constraintGreaterThanOrEqualToAnchor:self.buttonContainerView.leadingAnchor];
+    self.leadingButtonConstraintTop =
+        [self.leadingButton.topAnchor constraintEqualToAnchor:self.buttonContainerView.topAnchor];
+    self.leadingButtonConstraintTrailing = [self.trailingButton.trailingAnchor
+        constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
+    self.leadingButtonConstraintCenterY = [self.leadingButton.centerYAnchor
+        constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
+    self.leadingButtonConstraintTrailingWithTrailingButton = [self.leadingButton.trailingAnchor
+        constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
+    self.trailingButtonConstraintBottom = [self.trailingButton.bottomAnchor
+        constraintEqualToAnchor:self.buttonContainerView.bottomAnchor];
+    self.trailingButtonConstraintTop =
+        [self.trailingButton.topAnchor constraintEqualToAnchor:self.leadingButton.bottomAnchor
+                                                      constant:kButtonVerticalIntervalSpace];
+    self.trailingButtonConstraintTrailing = [self.trailingButton.trailingAnchor
+        constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +103,325 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    self.textLabelConstraintLeadingWithMargin = [self.textLabel.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
+
+    self.buttonContainerConstraintLeading = [self.buttonContainerView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
+    self.buttonContainerConstraintTrailing = [self.buttonContainerView.trailingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor constant:-kTrailingPadding];
```
clang-format suggested the following change:


```
-    self.imageViewConstraintLeading = [self.imageView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
-    self.imageViewConstraintTopSmall = [self.imageView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor constant:kTopPaddingSmall];
-    self.imageViewConstraintTopLarge = [self.imageView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor constant:kTopPaddingLarge];
-    self.imageViewConstraintBottom = [self.imageView.bottomAnchor constraintEqualToAnchor:self.layoutMarginsGuide.bottomAnchor constant:-kBottomPadding];
-
-    self.textLabelConstraintTop = [self.textLabel.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor constant:kTopPaddingLarge];
-    self.textLabelConstraintCenterY = [self.textLabel.centerYAnchor constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
-    self.textLabelConstraintTrailing = [self.textLabel.trailingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor constant:-kTrailingPadding];
-    self.textLabelConstraintLeadingWithImage = [self.textLabel.leadingAnchor constraintEqualToAnchor:self.imageView.trailingAnchor constant:kSpaceBetweenIconImageAndTextLabel];
-    self.textLabelConstraintLeadingWithMargin = [self.textLabel.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
-
-    self.buttonContainerConstraintLeading = [self.buttonContainerView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor constant:kLeadingPadding];
-    self.buttonContainerConstraintTrailing = [self.buttonContainerView.trailingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor constant:-kTrailingPadding];
-    self.buttonContainerConstraintBottom = [self.buttonContainerView.bottomAnchor constraintEqualToAnchor:self.layoutMarginsGuide.bottomAnchor constant:-kBottomPadding];
-    self.buttonContainerConstraintLeadingWithTextLabel = [self.buttonContainerView.leadingAnchor constraintEqualToAnchor:self.textLabel.trailingAnchor constant:kHorizontalSpaceBetweenTextLabelAndButton];
+    self.imageViewConstraintLeading =
+        [self.imageView.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor
+                                                     constant:kLeadingPadding];
+    self.imageViewConstraintTopSmall =
+        [self.imageView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor
+                                                 constant:kTopPaddingSmall];
+    self.imageViewConstraintTopLarge =
+        [self.imageView.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor
+                                                 constant:kTopPaddingLarge];
+    self.imageViewConstraintBottom =
+        [self.imageView.bottomAnchor constraintEqualToAnchor:self.layoutMarginsGuide.bottomAnchor
+                                                    constant:-kBottomPadding];
+
+    self.textLabelConstraintTop =
+        [self.textLabel.topAnchor constraintEqualToAnchor:self.layoutMarginsGuide.topAnchor
+                                                 constant:kTopPaddingLarge];
+    self.textLabelConstraintCenterY = [self.textLabel.centerYAnchor
+        constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
+    self.textLabelConstraintTrailing = [self.textLabel.trailingAnchor
+        constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor
+                       constant:-kTrailingPadding];
+    self.textLabelConstraintLeadingWithImage =
+        [self.textLabel.leadingAnchor constraintEqualToAnchor:self.imageView.trailingAnchor
+                                                     constant:kSpaceBetweenIconImageAndTextLabel];
+    self.textLabelConstraintLeadingWithMargin =
+        [self.textLabel.leadingAnchor constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor
+                                                     constant:kLeadingPadding];
+
+    self.buttonContainerConstraintLeading = [self.buttonContainerView.leadingAnchor
+        constraintEqualToAnchor:self.layoutMarginsGuide.leadingAnchor
+                       constant:kLeadingPadding];
+    self.buttonContainerConstraintTrailing = [self.buttonContainerView.trailingAnchor
+        constraintEqualToAnchor:self.layoutMarginsGuide.trailingAnchor
+                       constant:-kTrailingPadding];
+    self.buttonContainerConstraintBottom = [self.buttonContainerView.bottomAnchor
+        constraintEqualToAnchor:self.layoutMarginsGuide.bottomAnchor
+                       constant:-kBottomPadding];
+    self.buttonContainerConstraintLeadingWithTextLabel = [self.buttonContainerView.leadingAnchor
+        constraintEqualToAnchor:self.textLabel.trailingAnchor
+                       constant:kHorizontalSpaceBetweenTextLabelAndButton];
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +103,325 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+    self.leadingButtonConstraintTrailing = [self.leadingButton.trailingAnchor constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
+    self.leadingButtonConstraintCenterY = [self.leadingButton.centerYAnchor constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
+    self.leadingButtonConstraintTrailingWithTrailingButton = [self.leadingButton.trailingAnchor constraintEqualToAnchor:self.trailingButton.leadingAnchor constant:-kButtonHorizontalIntervalSpace];
+    self.trailingButtonConstraintBottom = [self.trailingButton.bottomAnchor constraintEqualToAnchor:self.buttonContainerView.bottomAnchor];
```
clang-format suggested the following change:


```
-    self.buttonContainerConstraintTopWithTextLabelGreater = [self.buttonContainerView.topAnchor constraintGreaterThanOrEqualToAnchor:self.textLabel.bottomAnchor constant:kVerticalSpaceBetweenButtonAndTextLabel];
-
-    self.leadingButtonConstraintLeading = [self.leadingButton.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.buttonContainerView.leadingAnchor];
-    self.leadingButtonConstraintTop = [self.leadingButton.topAnchor constraintEqualToAnchor:self.buttonContainerView.topAnchor];
-    self.leadingButtonConstraintTrailing = [self.leadingButton.trailingAnchor constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
-    self.leadingButtonConstraintCenterY = [self.leadingButton.centerYAnchor constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
-    self.leadingButtonConstraintTrailingWithTrailingButton = [self.leadingButton.trailingAnchor constraintEqualToAnchor:self.trailingButton.leadingAnchor constant:-kButtonHorizontalIntervalSpace];
-    self.trailingButtonConstraintBottom = [self.trailingButton.bottomAnchor constraintEqualToAnchor:self.buttonContainerView.bottomAnchor];
-    self.trailingButtonConstraintTop = [self.trailingButton.topAnchor constraintEqualToAnchor:self.leadingButton.bottomAnchor constant:kButtonVerticalIntervalSpace];
-    self.trailingButtonConstraintTrailing = [self.trailingButton.trailingAnchor constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
+    self.buttonContainerConstraintTopWithTextLabelGreater = [self.buttonContainerView.topAnchor
+        constraintGreaterThanOrEqualToAnchor:self.textLabel.bottomAnchor
+                                    constant:kVerticalSpaceBetweenButtonAndTextLabel];
+
+    self.leadingButtonConstraintLeading = [self.leadingButton.leadingAnchor
+        constraintGreaterThanOrEqualToAnchor:self.buttonContainerView.leadingAnchor];
+    self.leadingButtonConstraintTop =
+        [self.leadingButton.topAnchor constraintEqualToAnchor:self.buttonContainerView.topAnchor];
+    self.leadingButtonConstraintTrailing = [self.leadingButton.trailingAnchor
+        constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
+    self.leadingButtonConstraintCenterY = [self.leadingButton.centerYAnchor
+        constraintEqualToAnchor:self.buttonContainerView.centerYAnchor];
+    self.leadingButtonConstraintTrailingWithTrailingButton =
+        [self.leadingButton.trailingAnchor constraintEqualToAnchor:self.trailingButton.leadingAnchor
+                                                          constant:-kButtonHorizontalIntervalSpace];
+    self.trailingButtonConstraintBottom = [self.trailingButton.bottomAnchor
+        constraintEqualToAnchor:self.buttonContainerView.bottomAnchor];
+    self.trailingButtonConstraintTop =
+        [self.trailingButton.topAnchor constraintEqualToAnchor:self.leadingButton.bottomAnchor
+                                                      constant:kButtonVerticalIntervalSpace];
+    self.trailingButtonConstraintTrailing = [self.trailingButton.trailingAnchor
+        constraintEqualToAnchor:self.buttonContainerView.trailingAnchor];
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

jverkoey commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +106,371 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+
+#pragma mark - Layout helpers
+
+- (void)updateButtonsConstraintsWithLayoutMode:(MDCBannerViewLayoutMode)layoutMode {
```
Consider annotating this API as `NS_AVAILABLE_IOS(9_0)` so that you can wrap calls to it in `if (@available(iOS 9.0, *)) {` instead. This will let you delete the `if (@available(iOS 9.0, *)) {` check from this method's implementation.

fluxminds replied:
Good idea. Addressed in 0a1f2957e0cb3e9f3c11d9b7389b1f66a9b577c9.

jverkoey commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +106,371 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+  self.leadingButtonConstraintTrailingWithTrailingButton.active = NO;
+  self.trailingButtonConstraintBottom.active = NO;
+  self.trailingButtonConstraintTop.active = NO;
+  self.trailingButtonConstraintTrailing.active = NO;
```
I'm nervous about this method getting out of sync if/when new constraints are added. I unfortunately don't see an immediately obvious solution to this problem though. I think this is a good area for future improvement in terms of the maintenance complexity of this code.

jverkoey reviewed (commented):
I'm not convinced that this change in implementation has resulted in an increase in value overall. There is a lot of additional complexity due to managing so many constraints simply to get RTL support. Do you feel this code is easier to maintain than what was here previously?

Commit:
CRC changes: use NS_AVAILABLE_IOS to make internal method call cleaner.

fluxminds commented:
@jverkoey It is reasonable. I don't like the fact that many constraints need to be managed under this class, though I found it easy to debug with the current structure. I can try to refine/reduce the constraints used in this class (might not be many constrains because of the logic from design perspective). I can also fall back to full manual layout, which will lead to a little more calculation as what it used to be. The `layoutMode` will exist in both implementation as a helper property. What do you prefer?

material-automation commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +106,368 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+
+#pragma mark - Layout helpers
+
+- (void)updateButtonsConstraintsWithLayoutMode:(MDCBannerViewLayoutMode)layoutMode NS_AVAILABLE_IOS(9_0) {
```
clang-format suggested the following change:


```
-- (void)updateButtonsConstraintsWithLayoutMode:(MDCBannerViewLayoutMode)layoutMode NS_AVAILABLE_IOS(9_0) {
+- (void)updateButtonsConstraintsWithLayoutMode:(MDCBannerViewLayoutMode)layoutMode
+    NS_AVAILABLE_IOS(9_0) {
```

<!-- identifier: clang-format -->
<!-- File: components/Banner/examples/supplemental/MDCBannerView.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

jverkoey commented on components/Banner/examples/supplemental/MDCBannerView.m:
```diff
@@ -54,124 +106,368 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
+- (void)updateConstraintsWithLayoutMode:(MDCBannerViewLayoutMode)layoutMode {
+  [self deactivateAllConstraints];
+  // Set Elements
+  if (@available(iOS 9.0, *)) {
```
This method (and any others like it where all of the logic is conditional on the platform) can also make use of NS_AVAILABLE_IOS to remove a layer of logical indentation.

romoore replied:
+1 to marking available at compile time. Unfortunately, our catalogs/examples only seem to work with dynamic checks because we never explicitly depend on these classes when we compile/run.

fluxminds replied:
Addressed in 2bfba78355fdac2c698ea362166ec618d1a6aa5f

jverkoey commented:
> I don't like the fact that many constraints need to be managed under this class, though I found it easy to debug with the current structure.

Ah that's great to hear; out of curiosity what aspects of debugging did you find to be easier?

> I can try to refine/reduce the constraints used in this class (might not be many constrains because of the logic from design perspective).

If possible that might be helpful, though take care not to sacrifice clarity in doing so.

> I can also fall back to full manual layout, which will lead to a little more calculation as what it used to be. The layoutMode will exist in both implementation as a helper property. What do you prefer?

I had originally missed the deleted files because GitHub hid them from view, and I now see that the prior manual layout code was complex on roughly the same order of magnitude, so if you feel there's wins here of +1 debugging and +1 rtl support, then I'm supportive of the lateral movement.

I'd love to see a visualization of the constraints in each of the different modes. Can you add a screenshot of Xcode's visualization of the constraints for each mode to the PR description?

fluxminds commented:
@jverkoey 

> Ah that's great to hear; out of curiosity what aspects of debugging did you find to be easier?

I usually check if correct `layoutMode` is set, and then checking if constraints are set correctly from the view debugger.

> If possible that might be helpful, though take care not to sacrifice clarity in doing so.

It is a good suggestion. Actually something clarification was sacrificed to avoid introducing more constraints. But I am very willing to search for new constraint system from different perspective if possible. I will make a note and think of it. If available, I will send another PR as a complement. 

> I had originally missed the deleted files because GitHub hid them from view, and I now see that the prior manual layout code was complex on roughly the same order of magnitude, so if you feel there's wins here of +1 debugging and +1 rtl support, then I'm supportive of the lateral movement.

Though I don't like the fact that many constraints are included in this class now, the original layout code was a little unclear to read logic-wise, which was the reason as you mentioned. Another left thing I don't like is the `layoutMode` concept in this class, though it provides a layer to help adding specific constraints. I will keep working on it to see better alternative constraints can be used. 

I will add  a visualization of the constraints in each of the different modes to the description.

fluxminds committed:
[Banner] Update BannerTypicalUseExample with latest API changes (#6994)

Major changes:

- This PR updates the example for typical BannerView usage with latest BannerView API changes. 
- This PR add a separate example to each of three existing examples with no icon displayed on the banner view.
- This PR updated some constants in BannerTypicalUseExample. 


Screenshots Before Changes:
![Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 11 26](https://user-images.githubusercontent.com/8836258/55304790-2a2db700-541b-11e9-93b2-e402cb81b90c.png)
![Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 11 22](https://user-images.githubusercontent.com/8836258/55304794-2bf77a80-541b-11e9-8e4d-03b84cfeecb9.png)

Screenshots After Changes:
![Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 08 09](https://user-images.githubusercontent.com/8836258/55304800-3580e280-541b-11e9-979f-d0ced6529ffb.png)
![Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 08 06](https://user-images.githubusercontent.com/8836258/55304802-37e33c80-541b-11e9-970b-32c256857153.png)
![Simulator Screen Shot - iPhone 7 - 2019-04-01 at 01 08 05](https://user-images.githubusercontent.com/8836258/55304804-39ad0000-541b-11e9-84fe-6b258d2ae928.png)

Commit:
CRC changes: add NS_AVAILABLE_IOS
